crates/fuzzing/src/oracles/diff_wasmtime.rs

```diff
@@ -0,0 +1,184 @@
+
+        // Use the same allocation strategy between the two configs.
+        //
+        // Ideally this wouldn't be necessary, but if the lhs is using ondemand
+        // and the rhs is using the pooling allocator (or vice versa), then
+        // the module may have been generated in such a way that is incompatible
+        // with the other allocation strategy.
+        //
+        // We can remove this in the future when it's possible to access the
+        // fields of `wasm_smith::Module` to constrain the pooling allocator
+        // based on what was actually generated.
+        new_config.wasmtime.strategy = original_config.wasmtime.strategy.clone();
+        if let InstanceAllocationStrategy::Pooling { .. } = &new_config.wasmtime.strategy {
+            // Also use the same memory configuration when using the pooling allocator
+            new_config.wasmtime.memory_config = original_config.wasmtime.memory_config.clone();
+        }
+
+        WasmtimeEngine::new(&new_config)
+    }
+}
+
+impl DiffEngine for WasmtimeEngine {
+    fn instantiate(&self, wasm: &[u8]) -> Result<Box<dyn DiffInstance>> {
+        let mut store = self.config.to_store();
+        let module = compile_module(store.engine(), wasm, true, &self.config)
+            .context("unable to compile module in wasmtime")?;
+        let instance = instantiate_with_dummy(&mut store, &module)
+            .context("unable to instantiate module in wasmtime")?;
+        let instance = WasmtimeInstance {
+            store: RefCell::new(store),
+            instance,
+        };
+        Ok(Box::new(instance))
+    }
+}
+
+/// A wrapper around a Wasmtime instance.
+///
+/// The Wasmtime engine constructs a new store and compiles an instance of a
+/// Wasm module. The store is hidden in a [`RefCell`] so that we can hash, which
+/// does not modify the [`Store`] even though the API makes it appear so.
+struct WasmtimeInstance {
+    store: RefCell<Store<StoreLimits>>,
+    instance: Instance,
+}
+
+impl DiffInstance for WasmtimeInstance {
+    fn name(&self) -> &'static str {
+        "wasmtime"
+    }
+
+    fn evaluate(&mut self, function_name: &str, arguments: &[DiffValue]) -> Result<Vec<DiffValue>> {
+        let arguments: Vec<_> = arguments.iter().map(Val::from).collect();
+
+        let function = self
+            .instance
+            .get_func(self.store.borrow_mut().deref_mut(), function_name)
+            .expect("unable to access exported function");
+        let ty = function.ty(self.store.borrow().deref());
+        let mut results = vec![Val::I32(0); ty.results().len()];
+        function.call(
+            self.store.borrow_mut().deref_mut(),
+            &arguments,
+            &mut results,
+        )?;
+
+        let results = results.into_iter().map(Val::into).collect();
+        Ok(results)
+    }
+
+    fn is_hashable(&self) -> bool {
+        true
+    }
+
+    fn hash(&self, state: &mut std::collections::hash_map::DefaultHasher) -> Result<()> {
```

Comment:
Could this be `&mut self` to avoid the `RefCell` needed for Wasmtime?

Reply:
It could... but actually it feels like the Wasmtime API shouldn't require a mutable store when we're only reading values.

Reply:
Wasmtime has its own internal reasons for this, namely that extracting items from an instance is done lazily and may insert items into the store. There's other possibilities like reading an externref global could trigger a gc or similar situations like that. 

Overall I think it's best to stick with `&mut` here.